Parallelize CLAMP_DEVICE for each AMDGPU_TARGET

aaronenyeshi opened the pull request:
This change will make parallel the CLAMP_DEVICE calls for each AMDGPU_TARGET. With this patch, multiple CLAMP_DEVICE processes can run in parallel the major link-time tools opt and llc. Each clamp_device call should be target specific and thus not affect each other. There are significant performance gains on building projects like rocBLAS.

aaronenyeshi committed:
Parallelize CLAMP_DEVICE for each AMDGPU_TARGET

This change will make parallel the CLAMP_DEVICE calls for each AMDGPU_TARGET. With this patch, multiple CLAMP_DEVICE processes can run in parallel the major link-time tools opt and llc. Each clamp_device call should be target specific and thus not affect each other. There are significant performance gains on building projects like rocBLAS.

aaronenyeshi commented:
[ 88%] Building CXX object library/src/CMakeFiles/rocblas.dir/handle.cpp.o
[ 92%] Building CXX object library/src/CMakeFiles/rocblas.dir/rocblas_auxiliary.cpp.o
[ 96%] Building CXX object library/src/CMakeFiles/rocblas.dir/status.cpp.o
[100%] Linking CXX shared library librocblas.so
[100%] Built target rocblas

real    11m16.299s
user    14m49.281s
sys     0m15.842s

Seems we shave about 3m33s from rocBLAS build

scchan commented on lib/clamp-link.in:
```diff
@@ -406,19 +406,22 @@ if [ ${#LINK_KERNEL_ARGS[@]} != 0 ]; then
-    if [ $ret != 0 ]; then
-      exit $ret
-    fi
+    ret=$ret+$?;
```
potential race condition here?

scchan commented on lib/clamp-link.in:
```diff
@@ -406,19 +406,22 @@ if [ ${#LINK_KERNEL_ARGS[@]} != 0 ]; then
-    CLANG_OFFLOAD_BUNDLER_INPUTS_ARGS+=",$TEMP_DIR/kernel-$AMDGPU_TARGET.hsaco"
-    CLANG_OFFLOAD_BUNDLER_TARGETS_ARGS+=",hcc-amdgcn--amdhsa-$AMDGPU_TARGET"
+    CLANG_OFFLOAD_BUNDLER_INPUTS_ARGS+=",$TEMP_DIR/kernel-$AMDGPU_TARGET.hsaco";
+    CLANG_OFFLOAD_BUNDLER_TARGETS_ARGS+=",hcc-amdgcn--amdhsa-$AMDGPU_TARGET"; } &
```
I think we have a race condition here?

whchung commented:
@aaronenyeshi please help check if `xargs -P` allows better error handling and prevention of race condition

aaronenyeshi commented:
@whchung I think you can use the xargs -P implementation as well like this:
```
echo ${AMDGPU_TARGET_ARRAY[@]} | xargs -n 1 -P 4 sh -c ' $CLAMP_DEVICE $TEMP_DIR/kernel.bc $TEMP_DIR/kernel-$AMDGPU_TARGET.hsaco --amdgpu-target=$AMDGPU_TARGET; ret=$ret+$?; CLANG_OFFLOAD_BUNDLER_INPUTS_ARGS+=",$TEMP_DIR/kernel-$AMDGPU_TARGET.hsaco"; CLANG_OFFLOAD_BUNDLER_TARGETS_ARGS+=",hcc-amdgcn--amdhsa-$AMDGPU_TARGET"; '
```

But I don't think it would solve the race condition issue. 

@scchan I was wondering if using the bash `mkfifo ret` will solve race conditions using a named pipe fifo which will act as reading and writing to pipe and only accept one command on ret at a time.

scchan commented:
@aaronenyeshi could you put them into an array and then concat them after `wait`?

aaronenyeshi commented:
I'm still working on this. Seems that forked processes modify variables in their own fork process, and the parent process won't see it. So I will need to do some magic with PIDs to capture the exit code.

aaronenyeshi committed:
Fix race condition and error handling

Avoid race conditions by allowing assignments to be done sequentially. Provide better error handling by capturing the pids of forked processes, and then collecting their error codes when they are finished.

whchung commented on lib/clamp-link.in:
```diff
@@ -404,19 +404,24 @@ if [ ${#LINK_KERNEL_ARGS[@]} != 0 ]; then
+  # collect all the error codes from forked processes
+  # error handling
+  for pid in ${pids[*]}; do
+    wait $pid || ret=$((ret+$?))
```
@aaronenyeshi if i understand this correct, in case `any` target hits an error we'll abort the process, right?

aaronenyeshi replied:
Hmm, I don't think it will exit on error, rather it will keep error and exit when it detects it, same as before

aaronenyeshi replied:
Maybe I should put the if [$ret !=0] statement inside this loop

whchung replied:
i think it's better we wait on all subprocess to complete and determine whether to abort.

aaronenyeshi replied:
Yes, this implementation will satisfy that goal

scchan replied:
If a sub-process exited before wait is called, could you make sure that wait returns the correct return status of the terminated process and not 127?

---
man bash:

       wait [-n] [n ...]
              Wait for each specified child process and return its termination status.  Each n may be a process ID or a job specification; if a job spec is given, all processes in that job's pipeline are waited for.  If  n
              is  not  given, all currently active child processes are waited for, and the return status is zero.  If the -n option is supplied, wait waits for any job to terminate and returns its exit status.  If n speci‐
              fies a non-existent process or job, the return status is 127.  Otherwise, the return status is the exit status of the last process or job waited for.
---

aaronenyeshi replied:
Right, there is no guarantee that the sub-process won't exit before we call wait. One possible solution would be to put a sleep on the sub-process so that all the sub-processes could be invoked then we will wait on them before they wake up/potentially terminate.

whchung replied:
having a subprocess to sleep is messy. if we have to sleep i would think it'd be the master process (`clamp-link`) to do that. would this post offer some help?
https://stackoverflow.com/questions/356100/how-to-wait-in-bash-for-several-subprocesses-to-finish-and-return-exit-code-0

aaronenyeshi replied:
@scchan Actually i just tested it, and it seems that even if I sleep the main process longer than the sub-processes, the exit code is still maintained until wait is called. This was my test code:
```
  7   ret=0;
  8   declare -a AMDGPU_TARGET_ARRAY=('gfx803' 'gfx900')
  9   declare -a pids
 12  { sleep 3; echo --amdgpu-target=${AMDGPU_TARGET_ARRAY[0]}; exit 3; } &
 13  pids=("${pids[@]}" "$!")
 14  { sleep 1; echo --amdgpu-target=${AMDGPU_TARGET_ARRAY[1]}; exit 4; } &
 16  pids=("${pids[@]}" "$!")
 17  sleep 10;
 22
 23   for pid in ${pids[*]}; do
 24     wait $pid || ret=$((ret+$?))
 25   done
 26
 29   echo "ret: $ret"
 30   echo "pids: ${pids[@]}"
```
Output looks like this:
```
root@468e3189356e:~/hcc# ./test.sh
--amdgpu-target=gfx900
--amdgpu-target=gfx803
ret: 7
pids: 6791 6792
```

aaronenyeshi replied:
It seems that the sub-processes that we fork won't fall under `non-existent process or job, the return status is 127`. Seems for sub-processes even when finished before wait will stay active as a `zombie` process until the parent waits on it or is done

aaronenyeshi replied:
@whchung After experimentation, it should work without the sleeps
